### PR TITLE
Implement caching layer for validator table

### LINKED ISSUE #943: Implement caching layer for validator table
Extend what @taobun has written to make it fully featured, with moniker, description, etc. Make sure it's updated whenever there's a message to create validator or edit validator.

### PR DESCRIPTION
**Pull request by prin-r**
Fix #943

### CODE REVIEW

File: chain/app/db_app.go [hunk 1]
```diff
@@ -71,6 +71,24 @@ func (app *dbBandApp) InitChain(req abci.RequestInitChain) abci.ResponseInitChai
...
 					createMsg.ValidatorAddress,
 					createMsg.PubKey,
+					createMsg.Description.Moniker,
+					createMsg.Description.Identity,
+					createMsg.Description.Website,
+					createMsg.Description.Details,
+					createMsg.Commission.Rate.String(),
+					createMsg.Commission.MaxRate.String(),
+					createMsg.Commission.MaxChangeRate.String(),
+					sdk.NewCoin("uband", createMsg.MinSelfDelegation).String(),
```
Comment (sorawit) [inline chain/app/db_app.go]: Let's not hard-code `uband`. We can just store it add is. The frontend will decide if it wants to add the unit.

File: chain/db/Delegation.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package db
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (b *BandDB) AddDelegation(
```
Comment (sorawit) [inline chain/db/Delegation.go:8]: Please also handle `MsgDelegate` to make sure we keep track of delegations correctly.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/Delegation.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package db
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/staking"
+)
+
+func (b *BandDB) AddDelegation(
```
Comment (sorawit) [inline chain/db/Delegation.go:8]: This is still not correct. You will need to handle slashing, inflation, etc to make delegate works. Let's comment out all the code for now.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/validator.go [hunk 4]
```diff
@@ -88,3 +109,57 @@ func (b *BandDB) ClearOldVotes(currentHeight int64) error {
 	}
 	return nil
 }
+
+func (b *BandDB) GetValidator(validator Validator) (Validator, error) {
```
Comment (taobun) [inline chain/db/validator.go]: Change signature to GetValidator(validator sdk,ValAddress) (Validator, bool)

File: chain/db/validator.go [hunk 5]
```diff
@@ -88,3 +109,57 @@ func (b *BandDB) ClearOldVotes(currentHeight int64) error {
...
+	err := b.tx.First(&validator).Error
```
Comment (taobun) [inline chain/db/validator.go]: return true if found and false if error/not found

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/db/validator.go [hunk 6]
```diff
@@ -88,3 +109,63 @@ func (b *BandDB) ClearOldVotes(currentHeight int64) error {
 	}
 	return nil
 }
+
+func (b *BandDB) GetValidator(validator sdk.ValAddress) (Validator, bool) {
+	validatorSturct := Validator{OperatorAddress: validator.String()}
```
Comment (taobun) [inline chain/db/validator.go]: validatorStruct

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/db/validator.go [hunk 7]
```diff
@@ -88,3 +109,63 @@ func (b *BandDB) ClearOldVotes(currentHeight int64) error {
...
+	err := b.tx.First(&validatorSturct).Error
+	if err == nil {
```
Comment (taobun) [inline chain/db/validator.go]: return validatorStruct, err == nil

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/db/validator.go [hunk 8]
```diff
@@ -88,3 +109,58 @@ func (b *BandDB) ClearOldVotes(currentHeight int64) error {
...
+	return validatorStruct, err == nil
+}
+
+func (b *BandDB) handleMsgEditValidator(msg staking.MsgEditValidator) error {
+	validator, isFound := b.GetValidator(msg.ValidatorAddress)
+	if !isFound {
+		return fmt.Errorf(fmt.Sprintf("validator %s has not exist.", msg.ValidatorAddress.String()))
+	}
+
+	if msg.Description.Moniker != "[do-not-modify]" {
```
Comment (sorawit) [inline chain/db/validator.go]: Why do we type out string here? Let's use `DoNotModifyDesc` variable?

**Review by taobun — DISMISSED** [review]

File: chain/db/table.go [hunk 9]
```diff
@@ -14,12 +14,27 @@ type Event struct {
 	Name string
 }
 
-type Validator struct {
+type Delegation struct {
```
Comment (sorawit) [inline chain/db/table.go]: Let's comment out the whole table.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/validator.go (hunk 8 above)
Comment (taobun) [reply chain/db/validator.go]: https://github.com/cosmos/cosmos-sdk/blob/release/v0.37.8/x/staking/alias.go

**Review by taobun — DISMISSED** [review]

File: chain/db/validator.go [hunk 10]
```diff
@@ -88,3 +109,58 @@ func (b *BandDB) ClearOldVotes(currentHeight int64) error {
...
+func (b *BandDB) GetValidator(validator sdk.ValAddress) (Validator, bool) {
+	validatorStruct := Validator{OperatorAddress: validator.String()}
+	err := b.tx.First(&validatorStruct).Error
+	return validatorStruct, err == nil
+}
+
+func (b *BandDB) handleMsgEditValidator(msg staking.MsgEditValidator) error {
+	validator, isFound := b.GetValidator(msg.ValidatorAddress)
+	if !isFound {
+		return fmt.Errorf(fmt.Sprintf("validator %s has not exist.", msg.ValidatorAddress.String()))
```
Comment (sorawit) [inline chain/db/validator.go]: does not exist

File: chain/db/validator.go [hunk 11]
```diff
@@ -88,3 +109,58 @@ func (b *BandDB) ClearOldVotes(currentHeight int64) error {
...
+		validator.MinSelfDelegation = msg.MinSelfDelegation.ToDec().String()
+	}
+
+	return b.tx.Save(&validator).Error
+}
+
+func (b *BandDB) handleMsgCreateValidator(msg staking.MsgCreateValidator) error {
+	_, isFound := b.GetValidator(msg.ValidatorAddress)
+	if isFound {
+		return fmt.Errorf(fmt.Sprintf("validator %s has already exist.", msg.ValidatorAddress.String()))
```
Comment (sorawit) [inline chain/db/validator.go]: Why do we need to do this? I don't think it's even possible to not found.

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/validator.go (hunk 11 above)
Comment (taobun) [reply chain/db/validator.go]: to found ?

**Review by sorawit — APPROVED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**